cmd/gateway/s3x/gateway-s3x.go

```diff
@@ -95,26 +117,66 @@ func init() {
 
 func temxGatewayMain(ctx *cli.Context) {
 	minio.StartGateway(ctx, &TEMX{
-		HTTPAddr: ctx.String("info.http.endpoint"),
-		GRPCAddr: ctx.String("info.grpc.endpoint"),
-		DSPath:   ctx.String("ds.path"),
-		XAddr:    ctx.String("temporalx.endpoint"),
-		Insecure: ctx.Bool("temporalx.insecure"),
+		HTTPAddr:  ctx.String("info.http.endpoint"),
+		GRPCAddr:  ctx.String("info.grpc.endpoint"),
+		DSPath:    ctx.String("ds.path"),
+		DSType:    DSType(ctx.String("ds.type")),
+		CrdtTopic: ctx.String("ds.topic"),
+		XAddr:     ctx.String("temporalx.endpoint"),
+		Insecure:  ctx.Bool("temporalx.insecure"),
 	})
 }
 
-// newLedgerStore returns an instance of ledgerStore that uses badgerv2
-func (g *TEMX) newLedgerStore(dsPath string, dag pb.NodeAPIClient) (*ledgerStore, error) {
+// newLedgerStore returns an instance of ledgerStore
+func (g *TEMX) newLedgerStore(ctx context.Context, dag pb.NodeAPIClient) (*ledgerStore, error) {
+	switch g.DSType {
+	case DSTypeBadger:
+		return g.newBadgerLedgerStore(dag)
+	case DSTypeCrdt:
+		return g.newCrdtLedgerStore(ctx, dag)
+	}
+	return nil, fmt.Errorf(`data store type "%v" not supported`, g.DSType)
+}
+
+// newBadgerLedgerStore returns an instance of ledgerStore that uses badgerv2
+func (g *TEMX) newBadgerLedgerStore(dag pb.NodeAPIClient) (*ledgerStore, error) {
 	opts := badger.DefaultOptions
-	ds, err := badger.NewDatastore(dsPath, &opts)
+	ds, err := badger.NewDatastore(g.DSPath, &opts)
 	if err != nil {
 		return nil, err
 	}
 	return newLedgerStore(ds, dag)
 }
 
+// newCrdtLedgerStore returns an instance of ledgerStore that uses crdt and backed by badgerv2
+//
+func (g *TEMX) newCrdtLedgerStore(ctx context.Context, dag pb.NodeAPIClient) (*ledgerStore, error) {
+	store, err := badger.NewDatastore(g.DSPath, &badger.DefaultOptions)
+	if err != nil {
+		return nil, err
+	}
+	//from the doc: The broadcaster can be shut down by cancelling the given context. This must be done before Closing the crdt.Datastore, otherwise things may hang.
+	ctx, cancle := context.WithCancel(ctx)
+	pubsubBC, err := crdt.NewPubSubBroadcaster(ctx, dag, g.CrdtTopic)
+	if err != nil {
+		return nil, err
+	}
+	opts := crdt.DefaultOptions()
+	crdtds, err := crdt.New(store, datastore.NewKey("crdt"), dag, pubsubBC, opts)
+	if err != nil {
+		return nil, err
+	}
+	ls, err := newLedgerStore(crdtds, dag)
+	if err != nil {
+		return nil, err
+	}
+	ls.cleanup = append(ls.cleanup, cancle)
+	return ls, nil
+}
+
 // returns an instance of xObjects
 func (g *TEMX) getXObjects(creds auth.Credentials) (*xObjects, error) {
+	ctx := context.TODO()
```

Comment:
wouldn't it be better to use `context.Background` here?

Reply:
It was context.Background, but I think we should use a context that cancels when we nolonger need that xObject. `context.TODO()` is identical to `context.Background()` except it documents that the context should be replaced.